packages/onboard-helpers/src/hooks/useOnboardHelpers.ts

```diff
@@ -0,0 +1,356 @@
+              new Promise<never>((_, reject) => setTimeout(() => reject(new Error('timeout connect wallet')), 3000)),
+            ])
+            resolve(walletStates)
+          } catch (error) {
+            reject(error)
+          }
+        })
+
+        try {
+          const walletStates = await walletStatesPromise
+          if (!walletStates || walletStates?.length === 0) throw new Error('unable to connect')
+          walletState = walletStates[0]
+        } catch (error) {
+          // if failed to get walletState due to timeout, show connect modal.
+          updateStorageWalletName('')
+          ;[walletState] = await connect()
+        }
+      }
+
+      try {
+        if (!walletState) throw new Error('No wallet found')
+
+        updateStorageWalletName(walletState.label)
+        const { walletChainId, walletChainIdHex } = getWallet(walletState)
+
+        if (walletChainId && walletChainIdHex) {
+          let parsedChainId = +walletChainId
+
+          // if route and wallet chainId does not match, update wallet's chainId
+          if (+walletChainId !== routeChainId) {
+            const success = await setChain({ chainId: ethers.toQuantity(routeChainId) })
+            if (!success) throw new Error('reject network switch')
+            parsedChainId = routeChainId
+          }
+
+          // confirm wallet's chainId is updated
+          const { isNetworkSwitched } = await _isNetworkSwitched(routeChainId)
+
+          if (!isNetworkSwitched) {
+            const { networkId, isActiveNetwork } = getNetwork(parsedChainId)
+
+            if (networkId && isActiveNetwork) {
+              navigate(`${routeLocalePathname}/${networkId}/${routeFullPathname}`)
+              updateConnectApi([parsedChainId, true])
+            } else {
+              updateConnectState('failure', CONNECT_STAGE.SWITCH_NETWORK)
+            }
+            return
+          }
+          updateConnectApi([parsedChainId, true])
+        }
+      } catch (error) {
+        console.error(error)
+        updateStorageWalletName('')
+        updateConnectApi([routeChainId, false])
+      }
+    },
+    [
+      connect,
+      getNetwork,
+      navigate,
+      routeChainId,
+      routeFullPathname,
+      routeLocalePathname,
+      setChain,
+      updateConnectApi,
+      updateConnectState,
+      updateStorageWalletName,
+    ]
+  )
+
+  const handleDisconnectWallet = useCallback(
+    async (wallet: WalletState) => {
+      try {
+        updateStorageWalletName('')
+        await disconnect(wallet)
+        updateConnectApi([routeChainId, false])
+      } catch (error) {
+        console.error(error)
+      }
+    },
+    [disconnect, routeChainId, updateConnectApi, updateStorageWalletName]
+  )
+
+  const handleNetworkSwitch = useCallback(
+    async ([currChainId, newChainId]: ConnectOptions[CONNECT_STAGE.SWITCH_NETWORK]) => {
+      if (wallet && window?.ethereum) {
```

Comment:
Also, to reduce nesting, work the exceptions and return the result first, then proceed to the main handler
```suggestion
      if (!wallet || (window && !window.ethereum)) {
        updateConnectApi([newChainId, false])
        return
      }
```